pkg/traceql/ast_metrics.go

```diff
@@ -0,0 +1,376 @@
+	}
+}
+
+func (a *MetricsAggregate) observe(span Span) {
+	a.agg.Observe(span)
+}
+
+func (a *MetricsAggregate) observeExemplar(span Span) {
+	v, ts := a.exemplarFn(span)
+	a.agg.ObserveExemplar(span, v, ts)
+}
+
+func (a *MetricsAggregate) observeSeries(ss []*tempopb.TimeSeries) {
+	a.seriesAgg.Combine(ss)
+}
+
+func (a *MetricsAggregate) result() SeriesSet {
+	if a.agg != nil {
+		return a.agg.Series()
+	}
+
+	// In the frontend-version the results come from
+	// the job-level aggregator
+	return a.seriesAgg.Results()
+}
+
+func (a *MetricsAggregate) validate() error {
+	switch a.op {
+	case metricsAggregateCountOverTime:
+	case metricsAggregateMinOverTime:
+	case metricsAggregateMaxOverTime:
+	case metricsAggregateSumOverTime:
+	case metricsAggregateRate:
+	case metricsAggregateHistogramOverTime:
+		if len(a.by) >= maxGroupBys {
+			// We reserve a spot for the bucket so quantile has 1 less group by
+			return newUnsupportedError(fmt.Sprintf("metrics group by %v values", len(a.by)))
+		}
+	case metricsAggregateQuantileOverTime:
+		if len(a.by) >= maxGroupBys {
+			// We reserve a spot for the bucket so quantile has 1 less group by
+			return newUnsupportedError(fmt.Sprintf("metrics group by %v values", len(a.by)))
+		}
+		for _, q := range a.floats {
+			if q < 0 || q > 1 {
+				return fmt.Errorf("quantile must be between 0 and 1: %v", q)
+			}
+		}
+	default:
+		return newUnsupportedError(fmt.Sprintf("metrics aggregate operation (%v)", a.op))
+	}
+
+	if len(a.by) > maxGroupBys {
+		return newUnsupportedError(fmt.Sprintf("metrics group by %v values", len(a.by)))
+	}
+
+	return nil
+}
+
+var _ firstStageElement = (*MetricsAggregate)(nil)
+
+// secondStageElement represents operations that are performed
+// in the second stage metrics pipeline, such as topK/bottomK, etc.
+// these operations are performed on the metrics generated by the first stage.
+//
+// NOTE: find a batter name for this, maybe something like AggregateStage.
+// This stage operates on metrics generated by the first stage and performs aggregation on traceql metrics.
+// for now, calling it second stage is fine because it is the second stage in the pipeline.
+// and we already have MetricsAggregate which is the in the first stage that operates on spans
+// and produces metrics so we need to rename that to make things clear and avoid confusion.
+type secondStageElement interface {
+	Element
+	process(input SeriesSet) SeriesSet
+}
+
+type SecondStageOp int
+
+const (
+	OpTopK SecondStageOp = iota
+	OpBottomK
+)
+
+var errInvalidLimit = fmt.Errorf("limit must be greater than 0")
+
+func (op SecondStageOp) String() string {
+	switch op {
+	case OpTopK:
+		return "topk"
+	case OpBottomK:
+		return "bottomk"
+	}
+	return "unknown"
+}
+
+// TopKBottomK handles second stage topK/bottomK operations
+type TopKBottomK struct {
+	op    SecondStageOp
+	limit int
+}
+
+func newTopKBottomK(op SecondStageOp, limit int) *TopKBottomK {
```

Comment:
Debating on if all of the second stage ops need to be part of the same enum.  For example when we add another op, it would be possible to call `newTopKBottomK( somethingElse)` which shouldn't be valid.  Maybe this could be `newTopKBottomK(top bool)`